src/ProtoBuf.FSharp/Surrogates.fs

```diff
@@ -23,3 +24,60 @@ type Optional<'t> =
     static member RegisterIntoModel (model : RuntimeTypeModel) =
         let surrogateModelType = model.Add(typeof<Optional<'t>>, true)
         surrogateModelType.Name <- "Optional" + typeof<'t>.Name
+
+[<Struct; ProtoContract>]
+type ListSurrogate<'t> =
+    [<ProtoMember(1)>]
+    val mutable List : 't[]
+
+    static member op_Explicit (list: 't list) =
+        if obj.ReferenceEquals(list, null) || List.isEmpty list then
+            ListSurrogate(List = [| |])
+        else
+            ListSurrogate(List = List.toArray(list))
+
+    static member op_Explicit (sur: 't ListSurrogate) =
+        match sur.List with
+        | null | [| |] -> List.empty
+        | list when isApplicableTo typeof<'t> -> [ for el in list -> fixZeroValue el ]
+        | list -> List.ofArray list
+
+[<Struct; ProtoContract>]
+type SetSurrogate<'t when 't : comparison> =
+    [<ProtoMember(1)>]
+    val mutable List : 't[]
```

Comment:
Seems like it may support an IEnumerable collection here as the last fallback: https://github.com/protobuf-net/protobuf-net/blob/main/src/protobuf-net.Core/Serializers/RepeatedSerializer.cs#L452-L455.

NOTE: I believe it uses the TypeActivator to create a new instance of the collection (https://github.com/protobuf-net/protobuf-net/blob/main/src/protobuf-net.Core/Serializers/RepeatedSerializer.cs#L388). It means that we could create MutableSet wrapper types just for this purpose.

i.e. We don't necessarily need a surrogate either which may be interesting. If we are already creating record surrogates and such we just create MutableSet, MutableMap, Mutable... collections as the type of the record/unions.

They could look something like:

```
type MutableSet<'t> public () = 
    let mutable backingSet = Set.empty    
    member _.Result = backingSet // What op_Explicit or op_Implicit uses to convert back to the F# record/union.
    interface ICollection<'t> with
      member _.Add i = backingSet |> Set.add i
}
```

This could then go on the record or union surrogate as the type to use. We won't potentially need the List property as well which means a cleaner API for clients.

Reply:
There was some additional rationale behind using intermediate array: I was going to emit method like that
```F#
fixZeroValuesInplate : Array -> unit
```
with reflection to fix zero values in a more optimal way. For complex tree-like data structures like list, Set and Map overhead from one intermediate array is probably not that large (and if F# had custom methods to create Set-s and Map-s from sorted array it could work even better)

There's a [promise](https://github.com/protobuf-net/protobuf-net/blob/main/docs/3_0.md) of new API in the future:
```
new custom serializer API: there is a new API at the heart of protobuf-net that allows custom serializers to be defined in C# - both for messages and primitive values
```
which may solve a lot of problems. Although, there's still advantage in using surrogates: it allows to have something like `Set<int list>` --- that won't work now, but may be implemented with surrogates by adding one more wrapper onto internal `int list`

```
If we are already creating record surrogates and such we just create MutableSet, MutableMap, Mutable... collections as the type of the record/unions.
```
Yeah, I though about deep recursive data structures transformations --- that's really the only way that may solve problems with nulls and empty collections definitively --- not we can't fix something like:
```F#
type Record = {
    Field : struct (int[] * string)
}
```

Reply:
I would look forward to the new API to be frank. However note: I'm not looking to solve every single corner case with this library. If you are doing weird things with your contract types that are hard to express with a given .proto file, its probably not good to enable it this way either.

To be clear I would rather match the proto over full coverage of F# data (of course I want the latter too). The schema if generated should match what you would expect. For example

```

type TestRecordWithNestedCollections = {
    One: string list
    Two: Map<int, string list>
    Three: Set<string> 
}

```

Generates with your PR strange schema (I know it works)

```
message KeyValuePair_Int32_ListSurrogate_String {
   int32 Key = 1;
   ListSurrogate_String Value = 2;
}
message ListSurrogate_String {
   repeated string List = 1;
}
message MapSurrogate_Int32_ListSurrogate_String {
   repeated KeyValuePair_Int32_ListSurrogate_String List = 1;
}
message SetSurrogate_String {
   repeated string List = 1;
}
message TestRecordWithNestedCollections {
   repeated string One = 1;
   repeated string Three = 2;
   repeated KeyValuePair_Int32_ListSurrogate_String Two = 3;

```

Where really I just want a Map type.

```
message List_String_Envelope {
   repeated string List = 1;
}

message TestRecordWithNestedCollections {
   repeated string One = 1;
   map<int32, List_String_Envelope> Two = 2;
}
```

Ironically for this case I'm happy not to support it by default and recommend people create a ListStringEnvelope type they register into the model so that we can use map<int32, Contents> if its too hard to support all the cases.

NOTE: I'm happy with how Set<String> and List<String> turned out.

If we can get map looking like this then I might be able to tolerate the intermittent array. I'm not worried about the time overhead - I'm more worred about array allocation itself and the effect on the GC on a long running process. Lots of big arrays from unexpected messages is a thing that I've learnt not to do from experience due to the LOH - even if slower I'm thinking stability not performance.

Reply:
Which representation should maps use? I poked around a bit, and there are at least 2 working variants:
(simpler one; obvious problems --- intermediate dictionary and no zero value handling for keys)
```F#
[<Struct; ProtoContract>]
type MapSurrogate<'k, 'v when 'k : comparison> =
    [<ProtoMember(1); ProtoMap()>]
    val mutable Map : IDictionary<'k, 'v>

    static member op_Explicit (map: Map<'k, 'v>) : MapSurrogate<'k, 'v> =
        MapSurrogate(Map = map)

    static member op_Explicit (sur: MapSurrogate<'k, 'v>) : Map<'k, 'v> =
        match sur.Map with
        | null -> Map.empty
        | m when m.Count = 0 -> Map.empty
        | m when isApplicableTo typeof<'v> -> seq { for KeyValue(k, v) in m -> (k, fixZeroValue v) } |> Map.ofSeq
        | m -> seq { for KeyValue(k, v) in m -> (k, v) } |> Map.ofSeq
```

and
(quite ugly, but closer to your examples above)
```F#
[<AllowNullLiteral>]
type MutMap<'k, 'v when 'k : comparison> (map0 : Map<'k, 'v>) =
    let mutable map = map0

    new() = MutMap(Map.empty)
    member __.Map = map

    interface IEnumerable with
        member __.GetEnumerator() = (map :> IEnumerable).GetEnumerator()

    interface IEnumerable<KeyValuePair<'k, 'v>> with
        member __.GetEnumerator() = (map :> IEnumerable<_>).GetEnumerator()

    interface ICollection<KeyValuePair<'k, 'v>> with
        member __.Clear() = NotImplementedException "ICollection<_>.Clear" |> raise
        member __.Contains _ = NotImplementedException "ICollection<_>.Contains" |> raise
        member __.Remove _ = NotImplementedException "ICollection<_>.Remove" |> raise
        member __.CopyTo(_, _) = NotImplementedException "ICollection<_>.CopyTo" |> raise
        member __.Add _ = NotImplementedException "ICollection<_>.Add" |> raise

        member __.IsReadOnly = false
        member __.Count = map.Count

    interface IDictionary<'k, 'v> with
        member __.ContainsKey _ = NotImplementedException "IDictionary<_,_>.ContainsKey" |> raise
        member __.Remove (_ : 'k) = NotImplementedException "IDictionary<_,_>.Remove" |> raise

        member __.Keys = (map :> IDictionary<_, _>).Keys
        member __.Values = (map :> IDictionary<_, _>).Values
        member __.TryGetValue (k, refV) = map.TryGetValue(k, &refV)

        member __.Add (k, v) =
            map <- Map.add (fixZeroValue k) (fixZeroValue v) map

        member __.Item
            with get k = map.[k]
            and set k v = map <- Map.add (fixZeroValue k) (fixZeroValue v) map

[<Struct; ProtoContract>]
type MapSurrogate<'k, 'v when 'k : comparison> =
    [<ProtoMember(1); ProtoMap()>]
    val mutable Map : MutMap<'k, 'v>

    [<ProtoBeforeDeserialization>]
    member this.BeforeDeserialization () =
        this.Map <- MutMap<'k, 'v>()

    static member op_Explicit (map: Map<'k, 'v>) : MapSurrogate<'k, 'v> =
        if obj.ReferenceEquals(map, null) then
            MapSurrogate()
        else
            MapSurrogate(Map = MutMap<'k, 'v>(map))

    static member op_Explicit (sur: MapSurrogate<'k, 'v>) : Map<'k, 'v> =
        sur.Map.Map
```

Also, proto schema generation appears to be buggy: in the example above instead of
```
repeated KeyValuePair_Int32_ListSurrogate_String Two = 3;
```
in should be
```
MapSurrogate_Int32_ListSurrogate_String Two = 3;
```

Reply:
And now I realize that intermediate array surrogates actually use at least 2 intermediate arrays: first protobuf-net deserializes it into `System.Collection.Generic.List<_>` and then converts to array.

Reply:
To answer your questions implied or otherwise:

- "Which representation should maps use?" - In proto the map<_, _>. It expresses the intent.
- In code: The "ugly one" is better. I don't mind ugly when it isn't reflection emit code. At least it is self contained.

On the buggy proto generation that may be correct but that would be a separate issue to the protobuf-net repo. The schema generation is more to illustrate how Protobuf at the protocol level understands it and conversely what contract would work should other languages need to consumer our service (i.e. contract should be as clean as we can make it).

Reply:
> And now I realize that intermediate array surrogates actually use at least 2 intermediate arrays: first protobuf-net deserializes it into `System.Collection.Generic.List<_>` and then converts to array.

Yep. That's what I'm worried about. In the past I've found processes that use Protobuf do so due to the volume of data they are getting. Many people move to protobuf from JSON needing more performance or efficiency often with larger data as a cheap fix. Lots of large array allocations hurt the .NET runtime - hence Span<_>, ArraySegment, ArrayPool, Buffers, and other recent libraries to allow reuse of memory in a safe way. It is a real life case - I've had this exact problem with serialisers in the past and fragmented memory after a process has been running for awhile. Its an unexpected issue I would rather consumers avoid.

Reply:
After doing some work recently to complicate things I've also recently found that the Map type in Protobuf isn't that great.

https://developers.google.com/protocol-buffers/docs/proto3#maps

Key types can only be scalar (not including enum's). This is quite limiting in fact.

In hindsight the `repeated KV` form may be a better representation. Wasn't aware of this limitation.

Reply:
I'm not 100% sure that I remember it correctly, but I think protobuf-net should full abstract that: i.e. use map<> when possible and list of pairs otherwise